src/ray/raylet/worker_pool.cc

```diff
@@ -992,21 +992,36 @@ std::shared_ptr<WorkerInterface> WorkerPool::PopWorker(
       // Start a new worker process.
 
       if (task_spec.HasRuntimeEnv()) {
-        // create runtime env.
-        agent_manager_->CreateRuntimeEnv(
-            task_spec.JobId(), task_spec.SerializedRuntimeEnv(),
-            [this, start_worker_process_fn, &state, task_spec, runtime_env_hash](
-                bool successful, const std::string &serialized_runtime_env_context) {
-              if (successful) {
-                start_worker_process_fn(task_spec, state, {}, false, runtime_env_hash,
-                                        task_spec.SerializedRuntimeEnv(),
-                                        serialized_runtime_env_context);
-              } else {
-                RAY_LOG(ERROR) << "Creating runtime environment failed. The "
-                                  "corresponding task will be failed.";
-                runtime_env_setup_failed_callback_(task_spec.TaskId());
-              }
-            });
+        // Create runtime env.  If the env creation is already in progress on this
+        // node, skip this to prevent unnecessary CreateRuntimeEnv calls, which would
+        // unnecessarily start new worker processes.
+        auto it = runtime_env_statuses_.find(runtime_env_hash);
+        if (it == runtime_env_statuses_.end() || it->second == RuntimeEnvStatus::DONE) {
+          if (it == runtime_env_statuses_.end()) {
+            runtime_env_statuses_[runtime_env_hash] = RuntimeEnvStatus::PENDING;
+          }
+          agent_manager_->CreateRuntimeEnv(
+              task_spec.JobId(), task_spec.SerializedRuntimeEnv(),
+              [this, start_worker_process_fn, &state, task_spec, runtime_env_hash](
+                  bool successful, const std::string &serialized_runtime_env_context) {
+                runtime_env_statuses_[runtime_env_hash] = RuntimeEnvStatus::DONE;
+                if (successful) {
+                  start_worker_process_fn(task_spec, state, {}, false, runtime_env_hash,
+                                          task_spec.SerializedRuntimeEnv(),
+                                          serialized_runtime_env_context);
+                } else {
+                  RAY_LOG(ERROR) << "Creating runtime environment failed. The "
```

Comment:
I realize this was just copied/pasted from before, but can we include some more info in the error message (and also make it a warning since tasks can be retried by the end user)?

```
RAY_LOG(WARNING) << "Couldn't create a runtime environment for task " << task_spec.TaskId() << ". The runtime environment was " << task_spec.SerializedRuntimeEnv() << ".";
```